### PR TITLE
Persist only non empty properties of ConfigV6.Kibana and ConfigV7.Kibana in the open distro security index

### PR DESCRIPTION
**Pull request by vrozov**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java [hunk 1]
```diff
@@ -48,18 +49,57 @@
...
+import com.amazon.opendistroforelasticsearch.security.securityconf.impl.v7.ConfigV7;
 
 public class SecurityDynamicConfiguration<T> implements ToXContent {
     
     private static final TypeReference<HashMap<String,Object>> typeRefMSO = new TypeReference<HashMap<String,Object>>() {};
+    private static final String FOR_DISPLAY = "for_display";
+    public static final ToXContent.MapParams FOR_DISPLAY_PARAMS = new ToXContent.MapParams(Collections.singletonMap(FOR_DISPLAY, "true"));
+    private static final ObjectMapper objectMapper = createObjectMapper();
+
+    private static ObjectMapper createObjectMapper() {
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:80]: Can you move object mapper and associated object mapper settings to DefaultObjectMapper?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java [hunk 2]
```diff
@@ -48,18 +49,57 @@
...
+        final ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        objectMapper.enable(JsonParser.Feature.STRICT_DUPLICATE_DETECTION);
+        return registerModule(objectMapper);
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:84]: DefaultObjectMapper.objectMapper has both the properties set. (`NON_NULL` and `STRICT_DUPLICATE_DETECTION`). Can we not reuse it?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java [hunk 3]
```diff
@@ -132,8 +134,8 @@ public String toString() {
         }
     }
 
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java:137]: Please add unit tests for PATCH around this

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v6/ConfigV6.java [hunk 4]
```diff
@@ -81,8 +82,8 @@ public String toString() {
         }
     }
 
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v6/ConfigV6.java:85]: Please add unit tests for PATCH around this

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java [hunk 5]
```diff
@@ -48,18 +49,57 @@
...
 import com.amazon.opendistroforelasticsearch.security.securityconf.StaticDefinable;
-import com.amazon.opendistroforelasticsearch.security.securityconf.impl.v7.RoleV7;
+import com.amazon.opendistroforelasticsearch.security.securityconf.impl.v6.ConfigV6;
+import com.amazon.opendistroforelasticsearch.security.securityconf.impl.v7.ConfigV7;
 
 public class SecurityDynamicConfiguration<T> implements ToXContent {
     
     private static final TypeReference<HashMap<String,Object>> typeRefMSO = new TypeReference<HashMap<String,Object>>() {};
+    private static final String FOR_DISPLAY = "for_display";
+    public static final ToXContent.MapParams FOR_DISPLAY_PARAMS = new ToXContent.MapParams(Collections.singletonMap(FOR_DISPLAY, "true"));
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:77]: Could you add more information about `for_display` in PR/commit description ? Cannot see much information around this and what it means.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 1 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:80]: This `objectMapper` is created for `SecurityDynamicConfiguration` and is not intended for external usage.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 2 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:84]: No, it is a different instance of `objectMapper`.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 5 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:77]: `FOR_DISPLAY_PARAMS` changes `objectMapper` that is used to serialize `SecurityDynamicConfiguration` to JSON to include properties that will not be serialized by default `objectMapper` when they are `null` or empty.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java (hunk 3 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java:137]: Adding integration tests for PATCH is outside of scope for the PR, I'll add unit test for `ConfigV7.Kibana` serialization.

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java:137]: Okay can you add tests for serialization and deserialization then?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v6/ConfigV6.java (hunk 4 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v6/ConfigV6.java:85]: Please add relevant tests

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 2 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:84]: Why the same instance cannot be reused?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 1 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:80]: Is it possible to just expose whats needed similar to https://github.com/opendistro-for-elasticsearch/security/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/security/DefaultObjectMapper.java#L163

Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:80]: It is not currently supported and is not used outside of `SecurityDynamicConfiguration`.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 2 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:84]: It uses custom serializers for `ConfigV6.Kibana` and `ConfigV7.Kibana`.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java (hunk 3 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v7/ConfigV7.java:137]: I added unit tests for `ConfigV6.Kibana` and `ConfigV7.Kibana`.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v6/ConfigV6.java (hunk 4 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/v6/ConfigV6.java:85]: They are already added.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java (hunk 5 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:77]: Is "for_display" supposed to mean something?

Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/impl/SecurityDynamicConfiguration.java:77]: Yes, it means not for persistence. I can rename it to `TRANSIENT_PARAMS`. It is used to generate GET response in REST API.

**Review by sujithvm — DISMISSED** [review]

**Pull request closed without merging**